rig/machine_control/scp_connection.py

```diff
@@ -79,65 +124,169 @@ def send_scp(self, buffer_size, x, y, p, cmd, arg1=0, arg2=0, arg3=0,
             try:
-                # Try to receive the returned acknowledgement
                 ack = self.sock.recv(receive_length)
             except IOError:
-                # There was nothing to receive from the socket
-                continue
-
-            # Convert the possible returned packet into an SCP packet. If
-            # the sequence number matches the expected sequence number then
-            # the acknowledgement has been received.
-            scp = packets.SCPPacket.from_bytestring(ack, n_args=expected_args)
-
-            # Check that the CMD_RC isn't an error
-            if scp.cmd_rc in self.error_codes:
-                raise self.error_codes[scp.cmd_rc](
-                    "Packet with arguments: cmd={}, arg1={}, arg2={}, "
-                    "arg3={}; sent to core ({},{},{}) was bad.".format(
-                        cmd, arg1, arg2, arg3, x, y, p
+                # There wasn't a returned packet, we may spend quite some time
+                # here.
+                ack = None
+
+            # Process the received packet (if there is one)
+            if ack is not None:
+                # Extract the sequence number from the bytestring, iff possible
+                seq_bytes = ack[2 + consts.SDP_HEADER_LENGTH:
+                                2 + consts.SDP_HEADER_LENGTH + 4]
+                rc, seq = struct.unpack("<2H", seq_bytes)
+
+                # If the code is an error then we respond immediately
+                if rc in self.error_codes:
+                    raise self.error_codes[rc]
+
+                # Look up the sequence index of packet in the list of
+                # outstanding packets.  We may have already processed a
+                # response for this packet (indicating that the response was
+                # delayed and we retransmitted the initial message) in which
+                # case we can silently ignore the returned packet.
+                # XXX: There is a danger that a response was so delayed that we
+                # already reused the seq number... this is probably
+                # sufficiently unlikely that there is no problem.
+                outstanding = outstanding_packets.pop(seq, None)
+                if outstanding is not None:
+                    ack_packet = packets.SCPPacket.from_bytestring(
+                        ack, n_args=outstanding.expected_args
                     )
-                )
-
-            if scp.seq == self._seq:
-                # The packet is the acknowledgement.  Increment the
-                # sequence indicator and return the packet.
-                self._seq ^= 1
-                return scp
-
-        # The packet we transmitted wasn't acknowledged.
-        raise TimeoutError(
-            "Exceeded {} tries when trying to transmit packet.".format(
-                self.n_tries)
-        )
-
-    def read(self, buffer_size, x, y, p, address, length_bytes):
+                    outstanding.callback(ack_packet)
```

Comment:
We should switch to passing back the bytestring object, then it can be unpacked by the caller if necessary.